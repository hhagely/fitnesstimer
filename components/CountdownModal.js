import React, { Component } from 'react';
import { Modal, View, Text, StyleSheet, Button } from 'react-native';
import { Audio } from 'expo';

class CountdownModal extends Component {
	constructor(props) {
		super(props);
		this.state = this.initialState = {
			countdownTimeLeft: 0
		};
	}

	renderModal() {
		if (this.props.countdownTimeLeft !== 0) {
			return (
				<View
					style={[
						styles.container,
						{ alignItems: 'center' },
						this.props.countdownTimeLeft <= 3 &&
							this.props.countdownTimeLeft > 0 &&
							styles.countdownBackgroundAlt
					]}
				>
					<View>
						<Text style={{ fontSize: 32 }}>Your workout will begin in:</Text>
					</View>
					<View>
						<Text style={{ fontSize: 72 }}>{this.props.countdownTimeLeft}</Text>
					</View>
					<View style={styles.buttonContainer}>
						<Button onPress={this.props.cancelCountdown} title="Cancel" />
					</View>
				</View>
			);
		} else {
			return (
				<View
					style={[
						styles.container,
						{ alignItems: 'center' },
						styles.countdownBackgroundGo
					]}
				>
					<View>
						<Text style={{ fontSize: 108 }}>GO!</Text>
					</View>
				</View>
			);
		}
	}

	render() {
		return (
			<View>
				<Modal
					transparent={false}
					animationType="slide"
					visible={this.props.showModal}
					onRequestClose={this.props.cancelCountdown}
				>
					{this.renderModal()}
				</Modal>
			</View>
		);
	}
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		flexDirection: 'column',
		backgroundColor: '#fff',
		justifyContent: 'space-around'
	},
	contentContainer: {
		paddingTop: 30,
		paddingHorizontal: 12
	},
	buttonContainer: {
		width: 125
	},
	countdownBackgroundAlt: {
		backgroundColor: 'red'
	},
	countdownBackgroundGo: {
		backgroundColor: 'green'
	}
});

export default CountdownModal;
