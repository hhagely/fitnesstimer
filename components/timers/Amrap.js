import React, { Component } from 'react';
import { View, Button, StyleSheet } from 'react-native';
import { Audio } from 'expo';
import PropTypes from 'prop-types';

import TimeElapsed from '../TimeElapsed';
import CountdownModal from '../CountdownModal';

class AmrapTimer extends Component {
	constructor(props) {
		super(props);
		this.state = this.initialState = {
			timerSettings: this.props.timerSettings,
			showModal: false,
			isRunning: false,
			timeElapsed: 0,
			countdownTimeLeft: 0,
			startTime: 0
		};

		['update', 'reset', 'toggle', 'cancelCountdown'].forEach((method) => {
			this[method] = this[method].bind(this);
		});
	}

	toggleModal() {
		this.setState({ showModal: !this.state.showModal });
	}

	toggle() {
		this.setState(
			{
				isRunning: !this.state.isRunning
			},
			() => {
				const { timerSettings } = this.props;
				const { isRunning, timeElapsed } = this.state;
				if (isRunning && timeElapsed === 0) {
					if (timerSettings.countdown) {
						this.toggleModal();
						this.startCountdown();

						this.countdownTimerTimeout = setTimeout(() => {
							this.startTimer();
							this.toggleModal();
							clearInterval(this.countdownTimer);
							// add 1 second to the countdown duration so the countdown actually goes to zero
						}, timerSettings.countdownDuration * 1000 + 1000);
					} else {
						this.startTimer();
					}
				} else {
					isRunning && timeElapsed > 0
						? this.startTimer()
						: clearInterval(this.timer);
				}
			}
		);
	}

	reset() {
		clearInterval(this.timer);
		this.setState(this.initialState);
	}

	startTimer() {
		this.setState({ startTime: Date.now() });

		this.timer = setInterval(this.update, 1000);

		// this.endTimer = setTimeout(() => {
		// 	this.toggle();
		// }, this.props.timerSettings.timerDuration * 60 * 1000);
	}

	update() {
		const newTime = Date.now();
		const delta = newTime - this.state.startTime;

		let durationInMs = this.props.timerSettings.timerDuration * 60 * 1000; // 60k ms in 1 minute

		let tempElapsed = this.state.timeElapsed + delta;

		console.log('duration: ', this.props.timerSettings.timerDuration);
		console.log(`duration in ms: ${durationInMs}`);
		console.log(`temp elapsed: ${tempElapsed}`);

		this.setState(
			{
				startTime: newTime,
				timeElapsed: tempElapsed,
				timeLeft: durationInMs - tempElapsed
			},
			() => {
				if (-1000 <= this.state.timeLeft && this.state.timeLeft <= 0) {
					clearInterval(this.timer);
					this.toggle();
					this.setState(this.initialState);
				}
				console.log(`timeelapsed: ${this.state.timeElapsed}`);
				console.log(`time left: ${this.state.timeLeft}`);
			}
		);
	}

	startCountdown() {
		const { countdownDuration } = this.props.timerSettings;

		this.setState({ countdownTimeLeft: countdownDuration });

		this.countdownTimer = setInterval(
			() =>
				this.setState(
					{ countdownTimeLeft: this.state.countdownTimeLeft - 1 },
					() => {
						if (
							this.state.countdownTimeLeft <= 3 &&
							this.state.countdownTimeLeft > 0
						) {
							Audio.Sound.create(require('../../assets/sounds/Ping.mp3'), {
								shouldPlay: true
							});
						} else if (this.state.countdownTimeLeft === 0) {
							Audio.Sound.create(require('../../assets/sounds/Popcorn.mp3'), {
								shouldPlay: true
							});
						}
					}
				),
			1000
		);
	}

	cancelCountdown() {
		clearTimeout(this.countdownTimerTimeout);
		clearInterval(this.countdownTimer);

		this.setState({
			countdownTimeLeft: this.props.timerSettings.countdownDuration,
			isRunning: false,
			showModal: false
		});
	}

	render() {
		const { timerSettings } = this.props;
		const { timeElapsed, isRunning } = this.state;

		console.log(JSON.stringify(timerSettings));

		return (
			<View>
				<CountdownModal
					showModal={this.state.showModal}
					countdownTimeLeft={this.state.countdownTimeLeft}
					cancelCountdown={this.cancelCountdown}
				/>
				<TimeElapsed
					timeElapsed={timeElapsed}
					isRunning={isRunning}
					timerSettings={timerSettings}
				/>
				<View
					style={{
						flexDirection: 'row',
						justifyContent: 'space-around',
						alignItems: 'center'
					}}
				>
					<View style={styles.buttonContainer}>
						<Button
							onPress={this.toggle}
							title={isRunning ? 'Stop' : 'Start'}
						/>
					</View>

					<View style={styles.buttonContainer}>
						<Button
							onPress={this.reset}
							disabled={!isRunning && !timeElapsed}
							title="Reset"
						/>
					</View>
				</View>
			</View>
		);
	}
}

const styles = StyleSheet.create({
	buttonContainer: {
		width: 125
	}
});

AmrapTimer.propTypes = {
	timerSettings: PropTypes.shape({
		timerType: PropTypes.string,
		timerDuration: PropTypes.number,
		countdown: PropTypes.bool,
		countdownDuration: PropTypes.number,
		emomStyle: PropTypes.number
	})
};

export default AmrapTimer;
