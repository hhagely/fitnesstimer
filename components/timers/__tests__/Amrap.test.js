import { Button } from 'react-native';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';

import Amrap from '../Amrap';
import CountdownModal from '../../CountdownModal';

configure({ adapter: new Adapter() });

let wrapper;

let tempProps = {
	startCountdown: jest.fn(),
	timerSettings: {
		timerType: 'AMRAP',
		timerDuration: 10,
		countdown: true,
		countdownDuration: 10
	}
};

beforeEach(() => {
	wrapper = shallow(<Amrap {...tempProps} />);

	jest.useFakeTimers();
});

describe('<Amrap />', () => {
	it('renders the component', () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('calls startCountdown when start is pressed', () => {
		const startButton = wrapper.find(Button).at(0);

		startButton.simulate('press');

		wrapper.update();

		expect(wrapper.state('showModal')).toBe(true);
		expect(wrapper.state('isRunning')).toBe(true);

		jest.advanceTimersByTime(11000);

		expect(wrapper.state('showModal')).toBe(false);
		expect(clearInterval).toHaveBeenCalled();
	});

	it('calls the reset function when the button is pressed', () => {
		const startButton = wrapper.find(Button).at(0);
		const resetButton = wrapper.find(Button).at(1);

		startButton.simulate('press');

		wrapper.update();

		jest.advanceTimersByTime(11000);

		startButton.simulate('press');
		wrapper.update();
		resetButton.simulate('press');

		expect(clearInterval).toHaveBeenCalled();
		expect(wrapper.state()).toEqual(wrapper.instance().initialState);
	});

	it('calls endTimer when the timer is finished', () => {
		tempProps = {
			startCountdown: jest.fn(),
			timerSettings: {
				timerType: 'AMRAP',
				timerDuration: 1,
				countdown: false,
				countdownDuration: 10
			}
		};

		jest.spyOn(Amrap.prototype, 'endTimer');

		wrapper = shallow(<Amrap {...tempProps} />);

		const startButton = wrapper.find(Button).at(0);
		startButton.simulate('press');

		expect(Amrap.prototype.endTimer.mock.calls.length).toBe(0);

		jest.advanceTimersByTime(61000);

		// expect(Amrap.prototype.endTimer.mock.calls.length).toBe(1);
	});

	// it('cancels the countdown', () => {
	// 	const startButton = wrapper.find(Button).at(0);

	// 	startButton.simulate('press');

	// 	const modal = wrapper.find(CountdownModal).at(0);

	// 	console.log(modal.prop('showModal'));

	// 	wrapper.update();

	// 	wrapper.find(Button).map((b) => {
	// 		console.log('button title: ', b.prop('title'));
	// 	});
	// });
});
