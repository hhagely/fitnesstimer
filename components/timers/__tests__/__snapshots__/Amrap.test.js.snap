// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Amrap /> renders the component 1`] = `
<View>
  <CountdownModal
    cancelCountdown={[Function]}
    showModal={false}
  />
  <TimeElapsed
    isRunning={false}
    timeElapsed={0}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <View
      style={
        Object {
          "width": 125,
        }
      }
    >
      <Button
        onPress={[Function]}
        title="Start"
      />
    </View>
    <View
      style={
        Object {
          "width": 125,
        }
      }
    >
      <Button
        disabled={true}
        onPress={[Function]}
        title="Reset"
      />
    </View>
  </View>
</View>
`;
